### PR TITLE
Improve interval check method

### PR DESCRIPTION
**Pull request by conao3**
Those commit are commit in @zk-phi's fork ([repository](https://github.com/zk-phi/key-chord))

Accoding his [blog](https://qiita.com/zk_phi/items/e70bc4c69b5a4755edd6#key-chordel-%E3%81%AE%E8%AA%A4%E7%99%BA%E7%81%AB-%E3%82%92%E6%B8%9B%E3%82%89%E3%81%99), this idea is like below.

> So what I thought was to reduce the number of false inputs by
invoking when "one tempo delay from other keystrokes", instead of
just checking for "a short enough interval between two keystrokes".

### before
```
 \ カタカタ /      \ バン /     \ カタカタ /
h o g e h o g e      hj       h o g e h o g e . . .
----------------------------------------------------
                  |< 0.1s|
```

### after
```
 \ カタカタ /      \ バン /     \ カタカタ /
h o g e h o g e      hj       h o g e h o g e . . .
----------------------------------------------------
                 |< 0.15s|
              |> 0.1s||> 0.25s|
```

### CODE REVIEW

**Comment by irigone** [comment]
Why don't they accept the patch? Also, is the package unmaintained anymore?

**Comment by conao3** [comment]
What do you think about this PR, @tarsius? And are there any potential maintainers?

**Comment by tarsius** [comment]
> What do you think about this PR

Please put some more work into it before asking me to review. It appears that you are making some tweaks and then tweak the same things in the next commit or so. It is not at all obvious to me why you are doing that. Maybe there are reasons for doing that; if so, then please document them.

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+*.elc
+
```
Comment (tarsius) [inline .gitignore]: Please avoid adding a trailing empty line.

File: key-chord.el [hunk 2]
```diff
@@ -205,6 +205,9 @@
   "Max time delay between two press of the same key to be considered a key chord.
 This should normally be a little longer than `key-chord-two-keys-delay'.")
 
+(defvar key-chord-safety-interval 0.08
```
Comment (tarsius) [inline key-chord.el]: Later you replace that with two variables. Why not do that right away?

File: key-chord.el [hunk 3]
```diff
@@ -110,17 +110,6 @@
...
-;; write. Otherwise, if you are typing fast, two key intended to be separate
-;; letters might instead trig a chord.
-;; E.g. "uu" would be a good chord in spanish but not in finnish, and
-;; "hj" would be a good chord in english but not in swedish. 
-;;
-;; Don't rely solely on /usr/dict/words to find unusual combination.
-;; For example "cv" or "fg" can be quite common in certain kinds of
-;; programming. Grep your own texts to verify that a combination is unusual.
-;; And don't forget to check both permutations: "fg" and "gf".
-;;
```
Comment (tarsius) [inline key-chord.el]: Since the previous commit introduced the functional change that makes this tip obsolete (or at least I assume so), you should remove this tip in the same commit.

File: key-chord.el [hunk 4]
```diff
@@ -183,6 +172,8 @@
 
 ;; ########   History   ########################################
 ;;
+;; 0.7 (2017-06-29) zk-phi
+;; 	Add option `key-chord-safety-interval'
```
Comment (tarsius) [inline key-chord.el]: It makes little sense to release **in the middle** of a patch series. Also you should add the note about this new option in the commit that adds this new option. You can add a new history entry without releasing at the same time. Just use something like `Unreleased` instead of `0.7 (2017-06-29) zk-phi`.

File: key-chord.el [hunk 5]
```diff
@@ -189,14 +189,14 @@
...
+(defvar key-chord-two-keys-delay 0.15	; 0.05 or 0.1
   "Max time delay between two key press to be considered a key chord.")
 
-(defvar key-chord-one-key-delay 0.2	; 0.2 or 0.3 to avoid first autorepeat
+(defvar key-chord-one-key-delay 0.3	; 0.2 or 0.3 to avoid first autorepeat
   "Max time delay between two press of the same key to be considered a key chord.
 This should normally be a little longer than `key-chord-two-keys-delay'.")
 
-(defvar key-chord-safety-interval 0.08
+(defvar key-chord-safety-interval 0.15
```
Comment (tarsius) [inline key-chord.el]: Please explain why you are making these changes.

Comment (tarsius) [reply key-chord.el]: Also there is a typo in the commit message.

File: key-chord.el [hunk 6]
```diff
@@ -196,7 +196,7 @@
   "Max time delay between two press of the same key to be considered a key chord.
 This should normally be a little longer than `key-chord-two-keys-delay'.")
 
-(defvar key-chord-safety-interval 0.15
+(defvar key-chord-safety-interval 0.2
```
Comment (tarsius) [inline key-chord.el]: Why is that a separate commit?

File: key-chord.el [hunk 7]
```diff
@@ -173,7 +173,7 @@
 ;; ########   History   ########################################
 ;;
 ;; 0.7 (2017-06-29) zk-phi
-;; 	Add option `key-chord-safety-interval'
+;; 	Add option `key-chord-safety-interval-backward/forward'
```
Comment (tarsius) [inline key-chord.el]: You are adding **two** options and neither one is named `key-chord-safety-interval-backward/forward`.

File: key-chord.el [hunk 8]
```diff
@@ -199,7 +199,7 @@ This should normally be a little longer than `key-chord-two-keys-delay'.")
 (defvar key-chord-safety-interval-backward 0.1
   "Min time to distinguish a key chord and preceding inputs.")
 
-(defvar key-chord-safety-interval-forward 0.25
+(defvar key-chord-safety-interval-forward 0.35
```
Comment (tarsius) [inline key-chord.el]: Why is this being done in a separate commit?

File: key-chord.el [hunk 9]
```diff
@@ -240,10 +250,15 @@ pressed twice.
 			   (> (prefix-numeric-value arg) 0)
 			 (not key-chord-mode)))
   (cond (key-chord-mode
-	 (setq input-method-function 'key-chord-input-method)
+	 (setq input-method-function 'key-chord-input-method
+	       key-chord-timer-object
+	       (run-with-idle-timer key-chord-safety-interval t
+				    (lambda () (setq key-chord-idle-state t))))
```
Comment (tarsius) [inline key-chord.el]: It is better style to set just one variable per `setq` form. Stefan Monnier has said before that it was a mistake to allow setting multiple values and that the option to do this is only preserved for backward compatibility. (Same below.)

File: key-chord.el [hunk 10]
```diff
@@ -337,15 +353,17 @@ Please ignore that."
               (when (bound-and-true-p eldoc-mode)
                 (eldoc-pre-command-refresh-echo-area))
 
-	      (sit-for delay 0 'no-redisplay))
+	      (sit-for delay 'no-redisplay))
```
Comment (tarsius) [inline key-chord.el]: This seems to fix a bug and should be done in a separate commit.

File: key-chord.el [hunk 11]
```diff
@@ -337,15 +353,17 @@ Please ignore that."
...
 	    (progn
-	      (setq key-chord-last-unmatched nil)
+	      (setq key-chord-last-unmatched nil
+		    key-chord-idle-state     nil)
```
Comment (tarsius) [inline key-chord.el]: I am an "aligner" myself, but IMO you are overdoing it here and below.

**Review by tarsius — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (conao3) [reply .gitignore]: Thanks, this change is unneeded for this PR, I just remove this commit.

File: key-chord.el (hunk 3 above)
Comment (conao3) [reply key-chord.el]: OK, I do.

File: key-chord.el (hunk 7 above)
Comment (conao3) [reply key-chord.el]: I don't make a bump version, so this History entry was the whole disapear.

File: key-chord.el (hunk 8 above)
Comment (conao3) [reply key-chord.el]: OK, I fixup commits.

File: key-chord.el (hunk 9 above)
Comment (conao3) [reply key-chord.el]: Yes, I think so too. Fix done.

File: key-chord.el (hunk 10 above)
Comment (conao3) [reply key-chord.el]: OK, I make a separate commit.

File: key-chord.el (hunk 11 above)
Comment (conao3) [reply key-chord.el]: OK, fix it.

File: key-chord.el (hunk 2 above)
Comment (conao3) [reply key-chord.el]: OK, I fixup commit.

File: key-chord.el (hunk 4 above)
Comment (conao3) [reply key-chord.el]: Thanks, I don't make a bump version in this PR.

File: key-chord.el (hunk 6 above)
Comment (conao3) [reply key-chord.el]: OK, fixup commits.

File: key-chord.el (hunk 5 above)
Comment (conao3) [reply key-chord.el]: I don't know why he makes this change but this commit is disappeared after fixup commits.

**Comment by conao3** [comment]
Fix commit history and force pushed.

This PR was originally a separate fork by zk-phi, which he continued to develop as a fork, not to create a such PR.
It was my mistake to make it into a PR as it is. I apologize for the hassle and thanks for the good review.

**Comment by tarsius** [comment]
I actually wanted to try this package for a long time and now I finally have.

I also cleaned it. Please rebase this pr on top of that cleanup. Sorry for the additional work this causes.

> This PR was originally a separate fork by zk-phi, which he continued to develop as a fork, not to create a such PR.
> It was my mistake to make it into a PR as it is. I apologize for the hassle and thanks for the good review.

No problem. I didn't realize what was going on.

So I assume you tried this for a while. In your experience, how (much) doee it improve the user experience? I don't have enough experience using this package even *without* this change, so I intend to test both approaches for a while before making a decision.

**Comment by conao3** [comment]
rebase done.

To be honest, I'm not an enthusiastic fan of this package, but I thought zk-phi's suggestion made sense, so I suggested this PR.
(When he wrote the blog, upstream was already inactive and he was wondering where to send the patch)

It is reasonable to assume that the "key-chord decision" will be a little slower than the surrounding input speed, and I feel that this change has reduced the number of unintended command executions.

**Comment by tarsius** [comment]
Finally merged.

**Comment by conao3** [comment]
thanks!

**Comment by BlueDrink9** [comment]
@tarsius would you consider reverting this merge?

## New issues created by merge

This has introduced #6, #7 and #8. [ghost-of-freedom makes](https://github.com/emacsorphanage/key-chord/issues/8#issuecomment-1691475884) what I think is a good point, albeit brusquely:
Neither author nor merger uses this package much from what we can read here (correct me if this is wrong), whereas at least 11 people that do use it have been significantly enough disrupted by this PR to come here and mention it.

 ## No old issues solved by merge

This repo had sat for 4 years without accruing many issue reports. Two issues are referenced as being fixed by this PR, of which neither are actually solved by it. #5 had an existing correct solution found by the author (but they never closed the issue). #4 seems unrelated, presumably closed by mistake.

**Comment by tarsius** [comment]
I agree. Sorry about that.

**Comment by tarsius** [comment]
A similar change has been proposed in #12. Please help us test and review that, because otherwise my only options are to simply reject or to merge and risk a repeat of what happened in this case.
